python/rikai/spark/functions/vision.py

```diff
@@ -206,6 +215,152 @@ def video_to_images(
     ]
 
 
+video_metadata_schema = StructType(
+    [
+        StructField(
+            "error",
+            StructType(
+                [
+                    StructField("message", StringType(), True),
+                    StructField("stderr", StringType(), True),
+                    StructField("probe", StringType(), True),
+                ]
+            ),
+            True,
+        ),
+        StructField(
+            "data",
+            StructType(
+                [
+                    StructField("width", IntegerType(), True),
+                    StructField("height", IntegerType(), True),
+                    StructField("num_frames", IntegerType(), True),
+                    StructField("duration", FloatType(), True),
+                    StructField("bit_rate", IntegerType(), True),
+                    StructField("frame_rate", IntegerType(), True),
+                    StructField("codec", StringType(), True),
+                    StructField("size", IntegerType(), True),
+                ]
+            ),
+            True,
+        ),
+    ]
+)
+
+
+@udf(returnType=video_metadata_schema)
+def video_metadata(video: Union[str, VideoStream]) -> dict:
+    """Return useful video stream metadata about the given video
+
+    Parameters
+    ----------
+    video: str or VideoStream
+        The video uri or the Video object
+
+    Returns
+    -------
+    result: dict
+        { error: dict error info if ffprobe threw; otherwise None,
+          data: { width, height, num_frames, duration, bit_rate, frame_rate,
+                  codec, size}}
+
+    Notes
+    -----
+    The frame_rate is rounded to the nearest whole number of frames per sec
+
+    Examples
+    --------
+    The following returns the fps rounded to the nearest integer
+    ```
+    import rikai.spark.functions as RF
+    (spark.createDataFrame([(VideoStream(<uri>),)], ['video'])
+          .withColumn('meta', RF.video_metadata('video'))
+          .select('meta.data.frame_rate'))
+    ```
+    """
+    probe_result = _probe(video)
```

Comment:
This will raise exception, which will not be captured in the return?

Reply:
This will raise an exception if ffmpeg is not installed. In this case we just want to short circuit. The idea with the error handling is that each individual ffprobe call is to a commandline tool where the errors/exceptions often looks weird/opaque to python/scala.